### PR TITLE
Add endpoints for starting, stopping, and restarting all

### PR DESCRIPTION
**Pull request by albertshau**
cluster services, or specific cluster services, with all service
dependencies kept in mind. So if svcA depends on svcB, restarting
svcB will also restart svcA. Similarly, stopping svcB will first
stop svcA.

### CODE REVIEW

File: server/src/main/java/com/continuuity/loom/http/LoomClusterHandler.java [hunk 1]
```diff
@@ -644,6 +644,72 @@ public void putClusterConfig(HttpRequest request, HttpResponder responder,
...
+
+    try {
+      clusterService.requestServiceAction(clusterId, userId, action, service);
+      responder.sendStatus(HttpResponseStatus.OK);
+    } catch (MissingClusterException e) {
+      responder.sendError(HttpResponseStatus.NOT_FOUND, "Cluster " + clusterId + " not found.");
+    } catch (IllegalStateException e) {
+      responder.sendError(HttpResponseStatus.CONFLICT,
+                          "Cluster is not in a state where service actions can be performed.");
+    } catch (Exception e) {
```
Comment (poornachandra) [inline server/src/main/java/com/continuuity/loom/http/LoomClusterHandler.java:708]: Log the exception

File: server/src/main/java/com/continuuity/loom/scheduler/Actions.java [hunk 2]
```diff
@@ -103,24 +103,21 @@ public String toString() {
...
-                                       ProvisionerAction.CONFIGURE, ProvisionerAction.INITIALIZE,
-                                       ProvisionerAction.START),
-        ClusterAction.CLUSTER_CONFIGURE,
-                      ImmutableList.of(ProvisionerAction.BOOTSTRAP, ProvisionerAction.CONFIGURE),
-        ClusterAction.CLUSTER_CONFIGURE_WITH_RESTART,
-                      ImmutableList.of(ProvisionerAction.BOOTSTRAP, ProvisionerAction.STOP,
-                                       ProvisionerAction.CONFIGURE, ProvisionerAction.START));
-
+    this.actionOrder = ImmutableMap.<ClusterAction, List<ProvisionerAction>>builder()
+      .put(ClusterAction.CLUSTER_CREATE, ImmutableList.of(
```
Comment (poornachandra) [inline server/src/main/java/com/continuuity/loom/scheduler/Actions.java]: Can you add some spaces and newlines to make this map more readable?

File: server/src/main/java/com/continuuity/loom/scheduler/Actions.java [hunk 3]
```diff
@@ -131,28 +128,30 @@ public Actions() {
...
-        ClusterAction.CLUSTER_CONFIGURE_WITH_RESTART,
-                      ImmutableSet.of(
-                        // Start of a dependent service can happen only after its dependency has started
-                        new Dependency(ProvisionerAction.START, ProvisionerAction.START),
-                        // Stop of a dependent service can happen only before its dependency has stopped
-                        new Dependency(ProvisionerAction.STOP, ProvisionerAction.STOP, true)
-                      )
-      );
+    this.actionDependency = ImmutableMap.<ClusterAction, Set<Dependency>>builder()
+      .put(ClusterAction.CLUSTER_CREATE, ImmutableSet.of(
```
Comment (poornachandra) [inline server/src/main/java/com/continuuity/loom/scheduler/Actions.java]: More spaces and newlines to make this more readable.

File: server/src/main/java/com/continuuity/loom/scheduler/ClusterAction.java [hunk 4]
```diff
@@ -26,8 +29,13 @@
   CLUSTER_DELETE(Cluster.Status.INCOMPLETE),
   CLUSTER_CONFIGURE(Cluster.Status.INCONSISTENT),
   CLUSTER_CONFIGURE_WITH_RESTART(Cluster.Status.INCONSISTENT),
+  STOP_SERVICES(Cluster.Status.INCONSISTENT),
+  START_SERVICES(Cluster.Status.INCONSISTENT),
+  RESTART_SERVICES(Cluster.Status.INCONSISTENT),
   ADD_SERVICES(Cluster.Status.INCONSISTENT);
 
+  public static final Set<ClusterAction> SERVICE_ACTIONS = ImmutableSet.of(
```
Comment (poornachandra) [inline server/src/main/java/com/continuuity/loom/scheduler/ClusterAction.java]: Can you add a comment to say what a service action is?

File: server/src/main/java/com/continuuity/loom/scheduler/task/ClusterService.java [hunk 5]
```diff
@@ -117,6 +119,37 @@ public void requestClusterReconfigure(String clusterId, String userId, boolean r
...
+      }
+      if (!Cluster.Status.SERVICE_ACTIONABLE_STATES.contains(cluster.getStatus())) {
+        throw new IllegalStateException(
+          "cluster " + clusterId + " is not in a state where service actions can be performed");
+      }
+      JobId jobId = store.getNewJobId(clusterId);
+
+      ClusterJob job = new ClusterJob(jobId, action, service == null ? null : ImmutableSet.of(service), null);
+      job.setJobStatus(ClusterJob.Status.RUNNING);
+      cluster.addJob(job.getJobId());
```
Comment (poornachandra) [inline server/src/main/java/com/continuuity/loom/scheduler/task/ClusterService.java:141]: A long running cluster can have a lot of start stop jobs. We may need to prune the job list of the cluster.

File: server/src/main/java/com/continuuity/loom/scheduler/JobPlanner.java [hunk 6]
```diff
@@ -352,4 +363,23 @@ private boolean shouldPlanNode(Node node) {
   private boolean shouldPlanService(Service service) {
     return servicesToPlan == null || servicesToPlan.contains(service.getName());
   }
+
+  // if svc A depends on svc B depends on svc C, and if we're asked to restart svc B, we actually need to
```
Comment (poornachandra) [inline server/src/main/java/com/continuuity/loom/scheduler/JobPlanner.java]: "and if we're asked to restart svc B," -> "and if we're asked to restart svc C," ?

File: server/src/main/java/com/continuuity/loom/scheduler/task/ClusterService.java (hunk 5 above)
Comment (albertshau) [inline server/src/main/java/com/continuuity/loom/scheduler/task/ClusterService.java:141]: agreed, though I think this should be in a separate commit.  We have an issue open already (https://github.com/continuuity/loom/issues/18)

File: server/src/main/java/com/continuuity/loom/scheduler/JobPlanner.java (hunk 6 above)
Comment (albertshau) [inline server/src/main/java/com/continuuity/loom/scheduler/JobPlanner.java]: The comment is correct as is.  If we're asked to restart zookeeper, we need to restart hbase as well.  But the part about svc C is completely irrelevant and confusing so I will remove that part.

File: server/src/main/java/com/continuuity/loom/http/LoomClusterHandler.java (hunk 1 above)
Comment (albertshau) [inline server/src/main/java/com/continuuity/loom/http/LoomClusterHandler.java:708]: done

Comment (albertshau) [inline server/src/main/java/com/continuuity/loom/http/LoomClusterHandler.java:708]: the error message is also wrong, fixing it.

File: server/src/main/java/com/continuuity/loom/scheduler/Actions.java (hunk 3 above)
Comment (albertshau) [inline server/src/main/java/com/continuuity/loom/scheduler/Actions.java]: done

File: server/src/main/java/com/continuuity/loom/scheduler/Actions.java (hunk 2 above)
Comment (albertshau) [inline server/src/main/java/com/continuuity/loom/scheduler/Actions.java]: done

File: server/src/main/java/com/continuuity/loom/scheduler/ClusterAction.java (hunk 4 above)
Comment (albertshau) [inline server/src/main/java/com/continuuity/loom/scheduler/ClusterAction.java]: done
